jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/ForEachNodeInstance.java

```diff
@@ -153,6 +155,12 @@ public void internalTrigger(org.kie.api.runtime.process.NodeInstance fromm, Stri
             		String variableName = getForEachNode().getVariableName();
             		NodeInstance nodeInstance = (NodeInstance)
             		((NodeInstanceContainer) getNodeInstanceContainer()).getNodeInstance(getForEachSplitNode().getTo().getTo());
+
+            		// indirect to the real node if it is an async event node
+                    if (nodeInstance.getNode() instanceof AsyncEventNode && getNodeInstanceContainer() instanceof CompositeNodeInstance) {
+                        Node node = ((AsyncEventNode) nodeInstance.getNode()).getActualNode();
+                        nodeInstance = (NodeInstance) ((CompositeNodeInstance) getNodeInstanceContainer()).getNodeInstance(node, false);
```

Comment:
maybe an alternative approach would be to mark such nodes (entry node in foreach) as not async - like setting metadata on a node which would instruct to ignore asyncnode so it won't be wrapped. That would allow us to include that logic into the useAsync method, wdyt?

Reply:
@mswiderski  I cannot check the consequences of that. If we are talking about mutable instance another call in other place of the system could cause collateral problems. You know the code better, if those objects won't be used anywhere else then I don't have a problem implementing it that way.

Reply:
what you're doing here is essentially disabling async node, so better to already mark such node once (at parsing time) so you don't need to do that casting and unwrapping many times. Nodes are immutable, meaning they don't change per process instance and always stay the same as they are just the definition. So if you have a foreach on a node or subprocess it will always be used in that context

Reply:
Added like this.
https://github.com/kiegroup/jbpm/pull/1465/files#diff-9c56661f380ef8af8dabbd9268c7ea1aR437